clang/include/clang/AST/ASTConsumer.h

```diff
@@ -26,123 +26,132 @@ namespace clang {
-  /// definition. If it had an incomplete array type, its type will
-  /// have already been changed to an array of size 1. However, the
-  /// declaration remains a tentative definition and has not been
-  /// modified by the introduction of an implicit zero initializer.
-  virtual void CompleteTentativeDefinition(VarDecl *D) {}
-
-  /// CompleteExternalDeclaration - Callback invoked at the end of a translation
-  /// unit to notify the consumer that the given external declaration should be
-  /// completed.
-  virtual void CompleteExternalDeclaration(VarDecl *D) {}
-
-  /// Callback invoked when an MSInheritanceAttr has been attached to a
-  /// CXXRecordDecl.
-  virtual void AssignInheritanceModel(CXXRecordDecl *RD) {}
-
-  /// HandleCXXStaticMemberVarInstantiation - Tell the consumer that this
-  // variable has been instantiated.
-  virtual void HandleCXXStaticMemberVarInstantiation(VarDecl *D) {}
-
-  /// Callback involved at the end of a translation unit to
-  /// notify the consumer that a vtable for the given C++ class is
-  /// required.
-  ///
-  /// \param RD The class whose vtable was used.
-  virtual void HandleVTable(CXXRecordDecl *RD) {}
-
-  /// If the consumer is interested in entities getting modified after
-  /// their initial creation, it should return a pointer to
-  /// an ASTMutationListener here.
-  virtual ASTMutationListener *GetASTMutationListener() { return nullptr; }
-
-  /// If the consumer is interested in entities being deserialized from
-  /// AST files, it should return a pointer to a ASTDeserializationListener here
-  virtual ASTDeserializationListener *GetASTDeserializationListener() {
-    return nullptr;
-  }
-
-  /// PrintStats - If desired, print any statistics.
-  virtual void PrintStats() {}
-
-  /// This callback is called for each function if the Parser was
-  /// initialized with \c SkipFunctionBodies set to \c true.
-  ///
-  /// \return \c true if the function's body should be skipped. The function
-  /// body may be parsed anyway if it is needed (for instance, if it contains
-  /// the code completion point or is constexpr).
-  virtual bool shouldSkipFunctionBody(Decl *D) { return true; }
+  class TargetInfo;
+
+  /// ASTConsumer - This is an abstract interface that should be implemented by
+  /// clients that read ASTs.  This abstraction layer allows the client to be
+  /// independent of the AST producer (e.g. parser vs AST dump file reader,
+  /// etc).
+  class ASTConsumer {
+    /// Whether this AST consumer also requires information about
+    /// semantic analysis.
+    bool SemaConsumer = false;
+
+    friend class SemaConsumer;
+
+  public:
+    ASTConsumer() = default;
+
+    virtual ~ASTConsumer() {}
+
+    /// Initialize - This is called to initialize the consumer, providing the
+    /// ASTContext.
+    virtual void Initialize(ASTContext &Context) {}
+
+    /// Initialize - This is called to initialize the consumer, providing the
+    /// ASTContext.
```

Comment:
This should also say what `CodeGenTargetInfo` is used for, and that you need to make sure it's compatible with `Context`'s target info.